src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/BindToStreamBindingProvider.cs

```diff
@@ -0,0 +1,501 @@
+            }
+
+            var val = prop.GetValue(attribute);
+            var access = (FileAccess?)val;
+                        
+            return access;
+        }
+
+        private static void SetFileAccessFromAttribute(Attribute attribute, FileAccess access)
+        {
+            var prop = attribute.GetType().GetProperty("Access", BindingFlags.Public | BindingFlags.Instance);
+            // We already verified the type in GetFileAccessFromAttribute
+            prop.SetValue(attribute, access); 
+        }
+                
+        // As a binding, this is one per parameter, shared across each invocation instance.
+        private class ReadExactBinding : BindingBase<TAttribute>
+        {
+            private readonly BindToStreamBindingProvider<TAttribute> _parent;
+            private readonly Type _userType;
+            private readonly FileAccess _targetFileAccess;
+            private readonly Type _typeValueProvider;
+
+            public ReadExactBinding(
+                AttributeCloner<TAttribute> cloner,
+                ParameterDescriptor param,
+                BindToStreamBindingProvider<TAttribute> parent,
+                Type argHelper,
+                Type userType,
+                FileAccess targetFileAccess)
+                : base(cloner, param)
+            {
+                _parent = parent;
+                _userType = userType;
+                _targetFileAccess = targetFileAccess;
+                _typeValueProvider = argHelper;
+            }
+
+            protected override async Task<IValueProvider> BuildAsync(TAttribute attrResolved, ValueBindingContext context)
+            {
+                // set FileAccess beofre calling into the converter. Don't want converters to need to deal with a null FileAccess.
+                SetFileAccessFromAttribute(attrResolved, _targetFileAccess);
+
+                var patternMatcher = _parent._patternMatcher;
+                Func<object, object> builder = patternMatcher.TryGetConverterFunc(typeof(TAttribute), typeof(Stream));
+                Func<Stream> builder2 = () => (Stream)builder(attrResolved);
+
+
+                BaseValueProvider valueProvider = (BaseValueProvider)Activator.CreateInstance(_typeValueProvider);
+                await valueProvider.InitAsync(builder2, _userType);
+             
+                return valueProvider;
+            }
+        }
+
+        // The base IValueProvider. Handed  out per-instance
+        // This wraps the stream and coerces it to the user's parameter.
+        private abstract class BaseValueProvider : IValueBinder
+        {
+            private Stream _stream;
+            public Type Type { get; set; } // Impl IValueBinder
+
+            private string _invokeString;
+
+            // Helper to build the stream. This will only get invoked once and then cached as _stream. 
+            private Func<Stream> _streamBuilder;
+
+            object _userArg;
```

Comment:
can you clean up these declarations (order properly and remove all the unecessary spacing)? Also, I wonder if fxcop is running - should complain that you don't have "private" access modifier.